src/common/partitioning/src/lib.rs

```diff
@@ -0,0 +1,90 @@
+use std::{any::Any, sync::Arc};
+
+use common_error::DaftResult;
+use futures::stream::BoxStream;
+use serde::{Deserialize, Serialize};
+#[cfg(feature = "python")]
+use {
+    common_py_serde::{deserialize_py_object, serialize_py_object},
+    pyo3::PyObject,
+};
+
+/// Common trait interface for dataset partitioning, defined in this shared crate to avoid circular dependencies.
+/// Acts as a forward reference for concrete partition implementations. _(Specifically the `MicroPartition` type defined in `daft-micropartition`)_
+pub trait Partition: std::fmt::Debug + Send + Sync {
+    fn as_any(&self) -> &dyn std::any::Any;
```

Comment:
nit: → `as_any_ref`

Reply:
> nit: → `as_any_ref`

`as_any(&self) -> &dyn Any` is a pretty common pattern in rust.